src/models/losses.py

```diff
@@ -4,6 +4,327 @@
+    cond_err_g = binary_cross_entropy(cond_logits, real_labels)
+
+    uncond_logits = logits["fake"]["uncond"]
+    uncond_err_g = binary_cross_entropy(uncond_logits, real_labels)
+
+    loss_g = (
+        cond_err_g + uncond_err_g
+    )  # add up the conditional and unconditional losses
+    total_error_g += loss_g
+
+    local_incept_feat, global_incept_feat = inception_encoder(fake_imgs)
+    l1_w, l2_w, l1_s, l2_s = damsm_loss(
+        local_incept_feat,
+        global_incept_feat,
+        words_emb,
+        sent_emb,
+        match_labels,
+        cap_lens,
+        class_ids,
+        device,
+        const_dict,
+    )  # DAMSM Loss from attnGAN.
+
+    loss_damsm = lambda3 * (l1_w + l2_w + l1_s + l2_s)
+
+    total_error_g += loss_damsm
+
+    vgg_real = vgg_encoder(real_imgs)  # shape: (batch, 128, 128, 128)
+    vgg_fake = vgg_encoder(fake_imgs)  # shape: (batch, 128, 128, 128)
+
+    loss_per = mean_squared_error(vgg_real, vgg_fake)  # perceptual loss
+
+    total_error_g += loss_per / 2.0
+    return total_error_g
+
+
+def damsm_loss(
+    local_incept_feat: torch.Tensor,
+    global_incept_feat: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    device: Any,
+    const_dict: dict[Any, Any],
+) -> Any:
+    """Calculate the DAMSM loss from the attnGAN paper.
+
+    Args:
+        local_incept_feat: The local inception features. [shape: (batch, D, 17, 17)]
+
+        global_incept_feat: The global inception features. [shape: (batch, D)]
+
+        words_emb: The embeddings for all the words in the captions.
+
+        shape: (batch, D, max_caption_length)
+
+        sent_emb: The embeddings for the sentences. shape: (batch_size, D)
+
+        match_labels: Tensor of shape: (batch_size, 1).
+        This is of the form torch.tensor([0, 1, 2, ..., batch-1])
+
+        cap_lens: The length of the 'actual' captions in the batch [without padding]
+        shape: (batch_size, 1)
+
+        class_ids: The class ids for the instance. shape: (batch, 1)
+
+        device: The device to run the model on.
+
+        const_dict: The dictionary containing the constants.
+    """
+    batch_size = match_labels.size(0)
+    masks = (
+        []
+    )  # Mask mis-match samples, that come from the same class as the real sample ###
+
+    match_scores = []
+    cap_len_list = cap_lens.data.tolist()
+    gamma1 = const_dict["gamma1"]
+    gamma2 = const_dict["gamma2"]
+    gamma3 = const_dict["gamma3"]
+
+    for i in range(batch_size):
+        mask = (class_ids == class_ids[i]).type(torch.int)
+        mask[
+            i
+        ] = 0  # This ensures that "correct class" index is not included in the mask.
+        masks.append(mask.reshape(1, -1))  # shape: (1, batch)
+
+        numb_words = cap_len_list[i]
+        query_words = (
+            words_emb[i, :, :numb_words].unsqueeze(0).contiguous()
```

Comment:
i see.. yeah i guess felt that after transformation [transpose, unsqueeze, etc] the mapping from index to memory would not be the same so thought this might cause some issues... anyway will remove it for now..

Reply:
> mapping from index to memory would not be the same

If that's the case then I think you need to specify the desired shapes and if it doesn't make a difference then I would omit use of this function

> thought this might cause some issues

It surely might cause issues but in that case you will get a specific error about tensor not being contagious or having the wrong shape